File: packages/sdk/src/tests/unit/snapshotMigration0001.test.ts
import { SnapshotSchema } from '@towns-protocol/proto'
import { snapshotMigration0001 } from '../../migrations/snapshotMigration0001'
import { ethers } from 'ethers'
import { makeUniqueSpaceStreamId } from '../testUtils'
import { addressFromUserId, makeUniqueChannelStreamId, streamIdAsBytes } from '../../id'
import { check } from '@towns-protocol/dlog'
import { create } from '@bufbuild/protobuf'

// a no-op migration test for the initial snapshot, use as a template for new migrations
describe('snapshotMigration0001', () => {
    test('run migration', () => {
        const wallet = ethers.Wallet.createRandom()
        const userAddress = addressFromUserId(wallet.address)
        const spaceIdStr = makeUniqueSpaceStreamId()
        const spaceId = streamIdAsBytes(spaceIdStr)
        const channelIdStr = makeUniqueChannelStreamId(spaceIdStr)
        const channelId = streamIdAsBytes(channelIdStr)

        // members
        const badMemberSnap = create(SnapshotSchema, {
            members: {
                joined: [{ userAddress: userAddress }, { userAddress: userAddress }],
            },
        })
        const result = snapshotMigration0001(badMemberSnap)
        expect(result.members?.joined.length).toBe(1)

        // space channel payloads
        const badSpaceSnap = create(SnapshotSchema, {
            content: {
                case: 'spaceContent',
                value: {
                    channels: [{ channelId: channelId }, { channelId: channelId }],
                },
            },
        })
        const result2 = snapshotMigration0001(badSpaceSnap)
        check(result2.content.case === 'spaceContent')
        expect(result2.content?.value.channels.length).toBe(1)

        // user payload
        const badUserPayload = create(SnapshotSchema, {
            content: {
                case: 'userContent',
                value: {
                    memberships: [{ streamId: spaceId }, { streamId: spaceId }],
                },
            },
        })
        const result3 = snapshotMigration0001(badUserPayload)
        check(result3.content.case === 'userContent')
        expect(result3.content?.value.memberships.length).toBe(1)

        // user settings
        const badUserSettings = create(SnapshotSchema, {
            content: {
                case: 'userSettingsContent',
                value: {
                    fullyReadMarkers: [{ streamId: spaceId }, { streamId: spaceId }],
                },
            },
        })
        const result4 = snapshotMigration0001(badUserSettings)
        check(result4.content.case === 'userSettingsContent')
        expect(result4.content?.value.fullyReadMarkers.length).toBe(1)
    })
})

File: core/node/events/migrations/snapshot_migration_0001_test.go
package migrations

import (
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/towns-protocol/towns/core/node/base/test"
	"github.com/towns-protocol/towns/core/node/crypto"
	. "github.com/towns-protocol/towns/core/node/protocol"
	"github.com/towns-protocol/towns/core/node/testutils"
)

// nasty bug with the insert_sorted function, it was inserting an extra element at the end
// every insert, we need to remove duplicates

func TestSnapshotMigration0001(t *testing.T) {
	ctx := test.NewTestContext(t)
	userWallet, _ := crypto.NewWallet(ctx)
	spaceId := testutils.FakeStreamId(0x10) // events.STREAM_SPACE_BIN
	channelId := testutils.MakeChannelId(spaceId)

	// snaps have multiple member instances
	badMemberSnap := &Snapshot{
		Members: &MemberPayload_Snapshot{
			Joined: []*MemberPayload_Snapshot_Member{
				{
					UserAddress: userWallet.Address[:],
				},
				{
					UserAddress: userWallet.Address[:],
				},
			},
		},
	}
	// migrate
	snapshot_migration_0001(badMemberSnap)
	require.Equal(t, 1, len(badMemberSnap.Members.Joined))

	// space channel payloads
	badSpaceChannel := &Snapshot{
		Content: &Snapshot_SpaceContent{
			SpaceContent: &SpacePayload_Snapshot{
				Channels: []*SpacePayload_ChannelMetadata{
					{
						ChannelId: channelId[:],
					},
					{
						ChannelId: channelId[:],
					},
				},
			},
		},
	}
	snapshot_migration_0001(badSpaceChannel)
	require.Equal(t, 1, len(badSpaceChannel.GetSpaceContent().Channels))

	// user payload user membership
	badUserPayload := &Snapshot{
		Content: &Snapshot_UserContent{
			UserContent: &UserPayload_Snapshot{
				Memberships: []*UserPayload_UserMembership{
					{
						StreamId: spaceId[:],
					},
					{
						StreamId: spaceId[:],
					},
				},
			},
		},
	}
	snapshot_migration_0001(badUserPayload)
	require.Equal(t, 1, len(badUserPayload.GetUserContent().Memberships))

	// user settings fully read markers
	badUserSettings := &Snapshot{
		Content: &Snapshot_UserSettingsContent{
			UserSettingsContent: &UserSettingsPayload_Snapshot{
				FullyReadMarkers: []*UserSettingsPayload_FullyReadMarkers{
					{
						StreamId: channelId[:],
					},
					{
						StreamId: channelId[:],
					},
				},
			},
		},
	}
	snapshot_migration_0001(badUserSettings)
	require.Equal(t, 1, len(badUserSettings.GetUserSettingsContent().FullyReadMarkers))
}
